cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -52,12 +54,84 @@
 #include <memory>
 #include <optional>
 #include <type_traits>
+#include <unordered_set>
 #include <utility>
+#include <variant>
 
 namespace CUDF_EXPORT cudf {
 
 namespace experimental {
 
+/// Type identity type from C++20
+template <typename T>
+using type_identity_t = T;
+
+/**
+ * @brief Recursive template to apply type transformations sequentially. Transformations are
+ * applied first-to-last in the order specified.
+ *
+ * @tparam T Base type or initial type on which transformations are applied
+ * @tparam Rest Type transformations to apply
+ */
+template <typename T, template <typename> typename... Rest>
+struct dispatch_sequence;
+
+/// @copydoc dispatch_sequence
+template <typename T, template <typename> typename First, template <typename> typename... Rest>
+struct dispatch_sequence<T, First, Rest...> {
+  using type =
+    typename dispatch_sequence<First<T>, Rest...>::type;  ///< Resolved type after transformations
+};
+
+/// @copydoc dispatch_sequence
+template <typename T>
+struct dispatch_sequence<T> {
+  using type = T;  ///< The underlying type
+};
+
+/**
+ * @brief Helper alias for dispatch_sequence
+ */
+template <typename T, template <typename> typename... Rest>
+using dispatch_sequence_t = typename dispatch_sequence<T, Rest...>::type;
+
+/**
+ * @brief Void dispatcher helper
+ */
+template <bool B, typename T>
+using dispatch_void_conditional_t = std::conditional_t<B, void, T>;
+
+/**
+ * @brief Void dispatcher generator
+ */
+template <typename... Types>
+struct dispatch_void_conditional_generator {
+  /// The underlying type
+  template <typename T>
+  using type = dispatch_void_conditional_t<std::disjunction_v<std::is_same<T, Types>...>, T>;
+};
+
+/**
+ * @brief Returns `void` if it's a nested type
+ */
+template <typename T>
+using dispatch_void_if_nested_t =
+  dispatch_void_conditional_generator<id_to_type<type_id::STRUCT>,
+                                      id_to_type<type_id::LIST>>::type<T>;
+
+/**
+ * @brief Returns `void` if it's a compound type
+ *
+ */
+template <typename T>
+using dispatch_void_if_compound_t =
+  dispatch_void_conditional_generator<id_to_type<type_id::STRUCT>,
+                                      id_to_type<type_id::LIST>,
+                                      id_to_type<type_id::DECIMAL128>,
+                                      id_to_type<type_id::DECIMAL64>,
+                                      id_to_type<type_id::DECIMAL32>,
```

Comment:
This doesn't match the definition of `is_compound` from `cudf/utilities/traits.hpp`

Reply:
Good point. I initially suggested `compound` because I wanted to use an existing term rather than creating a new one. Ideally, the description would encompass compound types plus decimal types, but that's a bit verbose. How about `dispatch_void_if_complex_t` or `dispatch_void_if_composite_t`?

Reply:
Refined `compound` as `complex`